### DISCUSSION TITLE
RFC: Support arbitrary Cursor Positioning on VGA console?

### DISCUSSION THREAD

**Discussion by netzerohero** [picocomputer/community#77] [Q&A]
VGA (ver-0.6) console supports ansi colors (select-graphic-rendition SGR) and cursor forward and backward (CUF and CUB).

Consider similar support for arbitrary Cursor Positioning (via [ANSI escape codes](https://en.wikipedia.org/wiki/ANSI_escape_code)).
Cursor Position == (ESC [ n ; m H) == CUP == Moves the cursor to row n, column m.

This could provide additional capability for the VGA-console; e.g., eventual support within C- and BASIC- environments in a user-friendly way.

For example (in BASIC dialects): VTAB n : HTAB m (Applesoft) or LOCATE n,m (GW-BASIC).  The same feature would also be available in C via existing stdio printf(). This would open-up additional support to heritage BASIC code-bases and allow a c-programmer to use higher-level ansi standard stdio printfs without using lower-level calls to the VGA-graphics-pico directly (xreg calls).

**Comment by netzerohero** [comment]
**Analysis:**
Two approaches appear reasonable within rp6502 firmware:

1. Expand the **Console-Canvas-Type** to be similar to the **Text-Canvas-Type** in having a similar **Text-Config-Structure** containing a cursor x,y position.  Given static-scoping limits of the term_state structures, accessor-functions could then provide x,y cursor positions, since direct xreg access is not available.

![rp6502_Canvas_Screen_Types (copy)](https://github.com/picocomputer/community/assets/148929770/27569621-c9a8-4543-b172-6b659de3c792)


2. Expand function  rp6502:src/vga/term/term.c ::  **term_out_state_CSI()** to handle expanded CSI ansi-terminal-standard cases, including CUP - Cursor Position, and perhaps additional ansi=control-sequences as well (shown in yellow):
![AnsiTerminalCSIcommands (copy)](https://github.com/picocomputer/community/assets/148929770/8515e4d7-cbcd-4c3f-9a8f-e33e791c9804)

- Option-2 has the advantage of being "a standard driven" approach.
- Option-1 has the advantage of "growing into a solution" in that it aligns nicely to the existing Text-Canvas-Type (from C) and should low-level xreg access become available from EhBASIC.

**Comment by rumbledethumps** [comment]
If you want a mode 1 terminal emulator, write it on the 6502 side.

Adding mode 0 ANSI sequences has been low priority because there hasn't been a need. There's also a feature hold until https://github.com/picocomputer/rp6502/issues/51 gets fixed.

**Reply by netzerohero** [reply]
Ah, thanks for the insights.  For now I'm using minicom on the linux side.

**Reply by rumbledethumps** [reply]
Here's the supported ANSI sequences.
https://picocomputer.github.io/vga.html#terminal

**Comment by netzerohero** [comment]
RP6502 VGA (ver-0.12) adds additional CSI (Control Sequence Introducer) sequences to the VGA-console.
Nicely this includes: arbitrary Cursor Positioning (CUP): 

- Cursor Position == (ESC [ n ; m H) == CUP == Moves the cursor to row n, column m.

Note that beginning with RP6502 firmware versions (ver-0.10) requires upgrading the RP6502's picos from the original (RP2040-based) Picos to the newer (RP2350-based) Pico2s:

- RP6502 VGA now **requires** a Raspberry Pi Pico 2
- RP6502 RIA **requires** either a Pico 2 (non-wifi) or a Pico 2 W (wifi)

**Reply by rumbledethumps** [reply]
What will you do with your newfound power... the power of the cursor? Will you move gently, like a Terran scribe? Or will you strike boldly—leaping lines, bending code, carving paths through the document as a true warrior of the interface?

Choose wisely. Qapla’.

**Reply by netzerohero** [reply]
As soon as Mouser beams-over a pair of pico2 multi-core super-computer, er, glue-HW-upgrades, the power of the cursor you unleashed in FW ver-0.12 will join the full pre-IEEE-754 floating-point power within EHBasic to run unfettered on the 65c02 accessing heritage BASIC code-bases to implement 1st generation spreadsheet sources mimicking VisiCalc long-trapped within the [July-1982 issue of Softalk](https://archive.org/details/softalkv2n11jul1982/page/58/mode/1up?q=BASICALC) and the [Feb-1983 issue of REMark](https://www.rsp-italy.it/IT/Magazines/REMark/_contents/REMark_issue37_1983.pdf). Can't do that with today's integer-only cc65 compiling [Turbo-C MCALC](https://en.wikipedia.org/wiki/Turbo_C#Version_history) demo sources.  But a single 6502 in a stock A2 did it back-in-the-day as did the picocomputer's distant relative on a pair of Z80s in a [H89](https://en.wikipedia.org/wiki/Zenith_Z-89) running [SuperCalc](https://archive.org/details/eu_BYTE-1982-05_OCR/page/n227/mode/1up?view=theater). Alas, the original picos will be banished to the back of the parts drawer along with a long-forgotten slide-rule and a fondly thought of 1802.

Thank you for the upgrade.  Live long and prosper.

**Reply by rumbledethumps** [reply]
Fascinating. You can do that project entirely on device. llvm-mos does floats but doesn't yet have a math library. You probably only need a handful of functions that are trivial for AI. Which got me thinking how trivial it is to "type in" a program these days.
[BasiCalc.zip](https://github.com/user-attachments/files/21029589/BasiCalc.zip)

**Reply by netzerohero** [reply]
Here's the full original Apple-II BasiCalc source from 1982:
[BASICALC_A2_1982.txt](https://github.com/user-attachments/files/22191125/BASICALC_A2_1982.txt)

And[ here's my port.](https://github.com/orgs/picocomputer/discussions/152#discussion-8850978)
